Source/plrmsg.cpp

```diff
@@ -49,87 +62,69 @@ void plrmsg_delay(bool delay)
 	}
 
 	plrmsgTicks += SDL_GetTicks();
-	_plrmsg *pMsg = plr_msgs;
-	for (int i = 0; i < PMSG_COUNT; i++, pMsg++)
-		pMsg->time += plrmsgTicks;
+	for (PlayerMessage &message : Messages)
+		message.time += plrmsgTicks;
 }
 
-void ErrorPlrMsg(const char *pszMsg)
+void EventPlrMsg(string_view text, UiFlags style)
 {
-	_plrmsg *pMsg = &plr_msgs[plr_msg_slot];
-	plr_msg_slot = (plr_msg_slot + 1) & (PMSG_COUNT - 1);
-	pMsg->player = MAX_PLRS;
-	pMsg->time = SDL_GetTicks();
-	CopyUtf8(pMsg->str, pszMsg, sizeof(pMsg->str));
-}
+	PlayerMessage &message = GetNextMessage();
 
-size_t EventPlrMsg(const char *pszFmt, ...)
-{
-	_plrmsg *pMsg;
-	va_list va;
-
-	va_start(va, pszFmt);
-	pMsg = &plr_msgs[plr_msg_slot];
-	plr_msg_slot = (plr_msg_slot + 1) & (PMSG_COUNT - 1);
-	pMsg->player = MAX_PLRS;
-	pMsg->time = SDL_GetTicks();
-	vsprintf(pMsg->str, pszFmt, va);
-	va_end(va);
-	return strlen(pMsg->str);
+	message.style = style;
+	message.time = SDL_GetTicks();
+	message.text = std::string(text);
+	message.lineHeight = GetLineHeight(message.text, GameFont12) + 3;
 }
 
-void SendPlrMsg(int pnum, const char *pszStr)
+void SendPlrMsg(Player &player, string_view text)
 {
-	_plrmsg *pMsg = &plr_msgs[plr_msg_slot];
-	plr_msg_slot = (plr_msg_slot + 1) & (PMSG_COUNT - 1);
-	pMsg->player = pnum;
-	pMsg->time = SDL_GetTicks();
-	auto &player = Players[pnum];
-	assert(strlen(player._pName) < PLR_NAME_LEN);
-	assert(strlen(pszStr) < MAX_SEND_STR_LEN);
-	CopyUtf8(pMsg->str, fmt::format(_("{:s} (lvl {:d}): {:s}"), player._pName, player._pLevel, pszStr), sizeof(pMsg->str));
-}
+	PlayerMessage &message = GetNextMessage();
 
-void ClearPlrMsg()
-{
-	_plrmsg *pMsg = plr_msgs;
-	uint32_t tick = SDL_GetTicks();
+	std::string from = fmt::format(_("{:s} (lvl {:d}): "), player._pName, player._pLevel);
 
-	for (int i = 0; i < PMSG_COUNT; i++, pMsg++) {
-		if ((int)(tick - pMsg->time) > 10000)
-			pMsg->str[0] = '\0';
-	}
+	message.style = UiFlags::ColorWhite;
+	message.time = SDL_GetTicks();
+	message.text = from + std::string(text);
+	message.from = from;

```

Comment:
```suggestion
	message.from = std::string_view(message.text, from.size());
```

Could also remove the required space from the end of the `from` format string and add it as a joiner in the line above.

Reply:
The space should be exposed to translators.